### PR TITLE
0x6i/exception handling (WIP)

### PR DESCRIPTION
**Pull request by commjoen**
Hi all,
this is my first iteration for handling exceptions in iOS. I am not so sure about the content yet, please provide as much feedback as possible. 
With kind regards,
Jeroen

### CODE REVIEW

File: Document/0x06i-Testing-Code-Quality-and-Build-Settings.md [hunk 1]
```diff
@@ -216,23 +216,119 @@ In swift 3, using xCode 8, one can set Active Compilation Conditions setting in
 ### Testing Exception Handling
 
 #### Overview
+Exceptions can often occur when an application gets into a non-normal or erroneous state. 
+Testing exception handling is about reassuring that the application will handle the exception and get to a safe state without exposing any sensitive information at both the UI and the logging mechanisms used by the application.
 
--- TODO [Give an overview about the functionality "Testing Exception Handling" and it's potential weaknesses] --
+However, bare in mind that exception handling in objective-C is quite different than in Swift. Bridging the two concepts to one another in application that has both legacy objective-C code and Swift-code can be problematic.
```
Comment (Sjord) [inline Document/0x06i-Testing-Code-Quality-and-Build-Settings.md]: Shouldn't it be "bear in mind"?

File: Document/0x06i-Testing-Code-Quality-and-Build-Settings.md [hunk 2]
```diff
@@ -216,23 +216,119 @@ In swift 3, using xCode 8, one can set Active Compilation Conditions setting in
...
+* That any confidential information, such as keying material and/or authentication information is always wiped at the `@finally` blocks in case of a high risk application.
+* That `raise()` is only used in rare occassions when termination of the program without any further warning is required.
+* That `NSError` objects do not contain information that might leak any sensitive information.
+
+##### Static Analysis in Swift
+Here you can verify that:
+
+* The application uses a well-designed and unified scheme to handle errors.
+* The application doesn't expose sensitive information while handling errors in its UI or in its log-statements, but are still verbose enough to explain the issue to the user.
+* The throws are handled correctly: e.g.
```
Comment (Sjord) [inline Document/0x06i-Testing-Code-Quality-and-Build-Settings.md]: Is there an example missing here after the e.g.?

File: Document/0x06i-Testing-Code-Quality-and-Build-Settings.md [hunk 3]
```diff
@@ -216,23 +216,119 @@ In swift 3, using xCode 8, one can set Active Compilation Conditions setting in
...
+- If necessary, inform the user in an informative message to make him/her take appropriate action. The message itself should not leak sensitive information.
+- Not provide any information in logging mechanims used by the application.
+
+#### Remediation
+There are a few things a developer can do:
+- Ensure that the application use a well-designed and unified scheme to handle errors.
+- Make sure that all logging is removed or guarded as described in ["Testing for Debugging Code and Verbose Error Logging" for iOS]{TODO: whatlinkshouldbehere?}.
+- For Objective-C, in case of a high-risk application: create your own exception handler which cleans out any secret that should not be easily retrieved. The handler that can be set through `NSSetUncaughtExceptionHandler`.
+- When using Swift, make sure that you do not use `try!` unless you have made sure that there really cannot be any error in the method the throwing method that is being called.
+- When using Swift, make sure that the error does not propagate too far off.
```
Comment (Sjord) [inline Document/0x06i-Testing-Code-Quality-and-Build-Settings.md]: > too far off

Maybe "too far up the call stack"?

File: Document/0x06i-Testing-Code-Quality-and-Build-Settings.md (hunk 1 above)
Comment (commjoen) [inline Document/0x06i-Testing-Code-Quality-and-Build-Settings.md]: 👍

File: Document/0x06i-Testing-Code-Quality-and-Build-Settings.md (hunk 2 above)
Comment (commjoen) [inline Document/0x06i-Testing-Code-Quality-and-Build-Settings.md]: removed the sentence :).

File: Document/0x06i-Testing-Code-Quality-and-Build-Settings.md (hunk 3 above)
Comment (commjoen) [inline Document/0x06i-Testing-Code-Quality-and-Build-Settings.md]: That is true, but i rather say something like "- When using Swift, make sure that the error does not propagate too far off through intermediate methods." is that good as well?

**Comment by sushi2k** [comment]
Great test case @commjoen. Just a few minor things, I will merge the PR and make small changes (typo etc.)

**Comment by commjoen** [comment]
Thanks @sushi2k

**Pull request merged**